overlord/devicestate/devicemgr.go

```diff
@@ -679,6 +680,36 @@ func (m *DeviceManager) ensureCloudInitRestricted() error {
 			statusMsg = "failed to transition to done or error state after 5 minutes"
 		}
 
+		// we should always have a model if we are seeded and are not on classic
+		model, err := m.Model()
+		if err != nil {
+			return err
+		}
+
+		// For UC20, we want to always disable cloud-init after it has run on
+		// first boot unless we are in a "real cloud", i.e. not using NoCloud,
+		// or if we installed cloud-init configuration from the gadget
+		if model.Grade() != asserts.ModelGradeUnset {
+			// look for cloud.conf in the installed gadget snap
+			gadget := model.Gadget()
+			if gadget == "" {
+				// catastrophic error, we have a grade set but no gadget in the
+				// model?
+				return fmt.Errorf("internal error: model assertion has a grade of %s, but does not have a gadget", model.Grade())
+			}
+			gadgetInfo, err := snapstate.CurrentInfo(m.state, gadget)
+			if err != nil {
+				return err
+			}
```

Comment:
ideally this should use snapstate.GadgetInfo

Reply:
done